examples/ctrl-list.rs

```diff
@@ -0,0 +1,67 @@
+extern crate neli;
+use neli::consts::{CtrlAttr, CtrlCmd, GenlId, NlFamily, NlmF, Nlmsg};
+use neli::err::NlError;
+use neli::genl::Genlmsghdr;
+use neli::nl::Nlmsghdr;
+use neli::nlattr::Nlattr;
+use neli::socket::NlSocket;
+use neli::Nl;
+use neli::StreamReadBuffer;
+
+const GENL_VERSION: u8 = 2;
+
+// This example attempts to mimic the "genl ctrl list" command. For simplicity, it only outputs
+// the name and identifier of each generic netlink family.
+
+fn main() -> Result<(), NlError> {
+    let mut socket = NlSocket::connect(NlFamily::Generic, None, None, true)?;
+
+    let attrs: Vec<Nlattr<CtrlAttr>> = vec![];
+    let genlhdr = Genlmsghdr::new(CtrlCmd::Getfamily, GENL_VERSION, attrs)?;
+    let nlhdr = {
+        let len = None;
+        let nl_type = GenlId::Ctrl;
+        let flags = vec![NlmF::Request, NlmF::Dump];
+        let seq = None;
+        let pid = None;
+        let payload = genlhdr;
+        Nlmsghdr::new(len, nl_type, flags, seq, pid, payload)
+    };
+    socket.send_nl(nlhdr)?;
+
+    loop {
+        let response = socket.recv_nl::<GenlId, Genlmsghdr<CtrlCmd>>(None)?;
+
+        if let GenlId::UnrecognizedVariant(id) = response.nl_type {
+            match Nlmsg::from(id) {
+                // This example could be improved by reinterpreting the payload as an Nlmsgerr
+                // struct and printing the specific error encountered.
```

Comment:
I couldn't figure out an idiomatic way to do this with the library at the moment. If I missed something, let me know if there's a way to do this within neli and I can add the behavior mentioned in this comment.

Reply:
Ah! I see the misunderstanding. If you do the following, this might be easier.
```
let response = socket.recv_nl::<Nlmsg, Genlmsghdr<CtrlCmd>>(None)?;
match response.nl_type {
    Nlmsg::Error => {
        println!("An error occurred while retrieving available families");
        std::process::exit(1);
    },
    Nlmsg::Done => break,
    _ => (),
};
```
Because `Nlmsghdr` accepts a type parameter that can be satisfied with any type `T` as long as `T` is marked as compatible using `NlType`. `Nlmsg` is marked as `NlType`-compatible [here](https://github.com/jbaublitz/neli/blob/master/src/consts.rs#L497) so it can be passed as the first type parameter to `recv_nl` calls and return an `Nlmsghdr` with `nl_type` deserialized as the type `Nlmsg`. Does that make sense? I think this could be documented better at a high level so I'll make a note to do that.

Reply:
That makes sense. I'll make that change since it definitely simplifies this chunk of code.

The comment I left concerns the second type parameter more though. I was hoping to print out the error code within [`nlmsgerr`](https://github.com/jbaublitz/neli/blob/56bd6c7a180c780a4c7e13c3a70f9f9801237545/src/err.rs#L28), but that seems difficult since the buffer was already consumed as a `Genlmsghr`.

Reply:
After reviewing this further I think I see why you're doing it the way you're doing it - is it because you aren't sure whether a response is going to come back as `Nlmsg` type or `GenlId`? If that's the case, because you aren't actually using the `nl_type` field anywhere except to check for `Nlmsg::Done`, it may be worth doing what I suggest above. Typically what I do if I'm not sure what type is coming back and need to handle both scenarios is simply pass in `u16` as the type parameter (I've added the ability to use a raw `u16` as the type) and then try parsing it as both in whatever makes sense logically after it's been deserialized. It's less elegant but sometimes it makes more sense due to netlink being a heavily integer and C-targeted protocol. I'd like to make this less of a manual process but still am not sure how to make that better than what I just suggested at the moment.

Reply:
> After reviewing this further I think I see why you're doing it the way you're doing it - is it because you aren't sure whether a response is going to come back as Nlmsg type or GenlId?

Yes, that's exactly it. But not only are the types undetermined until response time, the payload is as well.

> If that's the case, because you aren't actually using the nl_type field anywhere except to check for Nlmsg::Done, it may be worth doing what I suggest above. Typically what I do if I'm not sure what type is coming back and need to handle both scenarios is simply pass in u16 as the type parameter (I've added the ability to use a raw u16 as the type) and then try parsing it as both in whatever makes sense logically after it's been deserialized.

Were you referring to the type or the payload here? I'm not aware of any way to parse a buffer as both for the later case. I agree it's trivial for the former (type) case.

> I'd like to make this less of a manual process but still am not sure how to make that better than what I just suggested at the moment.

Hm.. I think a Rust idiomatic way would be to define an enum that extends Nl.

```rust
#[derive(Nl)]
enum PossibleResponse {
  CtrlMsg(Genlmsg<CtrlCmd>),
  Nlmsgerr(Nlmsgerr<CtrlCmd>),
}
```

and then

```rust
let response = socket.recv_nl::<Nlmsg, PossibleResponse>(None)?;
```

I'll play with this method a bit.

Reply:
> Hm.. I think a Rust idiomatic way would be to define an enum that extends Nl.

Played around with it a bit but didn't get very far. We need a way to map message type ids to deserialization methods to do this, which is probably not within the scope of this pull request.

Reply:
I agree that this is outside the scope of this pull request but I made an issue #19 so we can discuss more there. This could be really beneficial for less manual work for users in parsing raw integers so I'm happy to move the discussion over there. I have some ideas for this and would love to bounce them off you before implementing. Thanks!